lib/rules/require-explicit-slots.js

```diff
@@ -0,0 +1,73 @@
+/**
+ * @author Mussin Benarbia
+ * See LICENSE file in root directory for full license.
+ */
+'use strict'
+
+const utils = require('../utils')
+
+module.exports = {
+  meta: {
+    type: 'problem',
+    docs: {
+      description: 'require slots to be explicitly defined with defineSlots',
+      categories: ['vue3-strongly-recommended'],
+      url: 'https://eslint.vuejs.org/rules/require-explicit-slots.html'
+    },
+    fixable: null,
+    schema: [],
+    messages: {
+      requireExplicitSlots:
+        'Slots must be explicitly defined with the defineSlots macro.',
+      alreadyDefinedSlot: 'Slot {{slotName}} is already defined.'
+    }
+  },
+  /** @param {RuleContext} context */
+  create(context) {
+    const slotsDefined = new Set()
+
+    return utils.compositingVisitors(
+      utils.defineScriptSetupVisitor(context, {
+        onDefineSlotsEnter(node) {
+          const typeArgument = node.typeArguments.params[0]
```

Comment:
Could you please add test cases that does not use `@typescript-eslint/parser`? Probably the `typeArguments ` property does not exist.

Reply:
I see! This is why some other rules actually define the `@typescript-eslint/parser` option in a separate object and then spread it into some tests. I was considering TS by default but that's a wrong assumption to make. Thank you for the suggestion, I'll add the test 👍

Reply:
Unfortunately I seem to get a parsing error if I don't use `@typescript-eslint/parser` 🤔 

```sh
1) require-explicit-slots
       valid

      <template>
        <div>
          <slot />
        </div>
      </template>
      <script setup lang="ts">
      defineSlots<{
        default(props: { msg: string }): any
      }>()
      </script>:

      AssertionError [ERR_ASSERTION]: A fatal parsing error occurred: Parsing error: Unexpected token :
      + expected - actual

      -false
      +true
```

All test cases fail with the same error, a colon `:` seems to get appended at the end of the script closing tag. Since the syntax we are looking for is TS, I thought we must require `@typescript-eslint/parser`, am I misunderstanding?

cc @FloEdelmann

Reply:
When not using the TS parser, you also can't use TS syntax, i.e. `lang="ts"` and type annotations like `: string` need to be removed.

Reply:
Makes sense, the thing I can't understand though is whether testing without TS is required in this case since both [defineSlots](https://vuejs.org/api/sfc-script-setup.html#defineslots) and the option api [slots](https://vuejs.org/api/options-rendering.html#slots) require TS in the first place.

If we disable the rule when the script does not have `lang="ts"`, as per @ota-meshi's suggestion below, then the tests can just assume that TS is being used, correct?

Reply:
Hmm true, `defineSlots`/`slots` don't make much sense without TypeScript. But I think they would not break the build without it, right? So I think the rule should still run in a JavaScript-only component.

If we agree that the rule doesn't run in a non-TS component, then a test should verify that this rule doesn't report any warnings for a non-TS component not defining `defineSlots`/`slots` but still using `<slot>`.